runtime-modules/content/src/lib.rs

```diff
@@ -0,0 +1,1142 @@
+    /// If set, updates the seasons of a series. Extend a series when length of seasons is
+    /// greater than previoulsy set. Last elements must all be 'Some' in that case.
+    /// Will truncate existing series when length of seasons is less than previously set.
+    seasons: Option<Vec<Option<SeasonParameters<VideoCategoryId, VideoId>>>>,
+    meta: Option<Vec<u8>>,
+}
+
+/// A season is an ordered list of videos (episodes).
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct Season<VideoId> {
+    episodes: Vec<VideoId>,
+}
+
+/// A series is an ordered list of seasons that belongs to a channel.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct Series<ChannelId, VideoId> {
+    in_channel: ChannelId,
+    seasons: Vec<Season<VideoId>>,
+}
+
+// The actor the caller/origin is trying to act as for Person creation and update and delete calls.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum PersonActor<MemberId, CuratorId> {
+    Member(MemberId),
+    Curator(CuratorId),
+}
+
+/// The authorized actor that may update or delete a Person.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum PersonController<MemberId> {
+    /// Do not use - Default value representing empty value
+    Nobody,
+    /// Member controls the person
+    Member(MemberId),
+    /// Any curator controls the person
+    Curators,
+}
+
+// Default trait implemented only because its used in Person which needs to implement a Default trait
+// since it is a StorageValue.
+impl<MemberId> Default for PersonController<MemberId> {
+    fn default() -> Self {
+        PersonController::Nobody
+    }
+}
+
+/// Information for Person being created.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct PersonCreationParameters {
+    /// Metadata for person.
+    meta: Vec<u8>,
+}
+
+/// Information for Persion being updated.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct PersonUpdateParameters {
+    /// Metadata to update person.
+    new_meta: Vec<u8>,
+}
+
+/// A Person represents a real person that may be associated with a video.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct Person<MemberId> {
+    /// Who can update or delete this person.
+    controlled_by: PersonController<MemberId>,
+    /// The number of videos this person appears in.
+    number_of_videos_person_involed_in: u32,
```

Comment:
Where is it stored what specific videos/seasons/series a person appears in? How this should be done depends on who manages creation and destruction of such relationships. One way is to allow content owners to add person to something. That scales well, and it prevents someone from spamming a person onto tons of content, like adding "Jesus" to everything. At the same time, do we then allow a person to be deleted from the directory if third parties added them to their content? I think that is OK. In that case, content can refer to dangling persons. In this case, presence of persons on content is just part of their metadata, and no reference counting is needed.

What do you think?

Reply:
Obviously I hadn't expanded on the implementation of `add_person_to_video` and `remove_person_from_video()` but that is indeed the permission model I was planning to go with.

With respect to deletion, yes I don't see a cleaner way to deal with constraint and resulting dangling references. Putting it in metadata both reduces state burden and code to keep track of references, and moves the burden on UI apps to deal with missing persons (similarly with missing videos from a playlist)

Reply:
Two additional thoughts come to mind:

1 - With the `Person` struct now being reduced to having a single field (PersonController), does it make sense to keep it as a struct an maintain the map:  `PersonById: PersonId => Person` or might as well just drop it and have a map: `PersonControlledBy: PersonId => PersonController`

2 - We have a similar situation with ChannelCategory and VideoCategory. Which are maintaining counters to prevent deletion of the categories if videos or channels are in them.. should I apply same treatment and let applications deal with dangling reference to category ids?

```
pub struct VideoCategory {
    /// The number of videos in this category.
    number_of_videos_in_category: u32,
}

pub struct ChannelCategory {
    /// The number of channels in this category.
    number_of_channels_in: u32,
}
```

Reply:
I would keep them, even if just with one field. Its semantically cleaner and we will possibly find reasons to expand later.

Reply:
2. Following up on this point.

I guess that is ok, otherwise people can block deletion of categories by having their content in there. This does however mean that people will have invalid category information, which I guess is OK, at least for now.

Reply:
Updated Person representation in runtime and added to video metadata in e19e1f677e2f3fd588ff7dd3d2446ca8b5538fc0
Added persons fields to Series and Season metadata in 8974e224017107a1fcb14217fcbbac9ea907c8a9

Reply:
removed counters in Category types in 35c00c0dc428d8c97d02e9f9123f1d002da846fe

Reply:
category assigned in metadata instead in a903a493d8f612c13c88dcfd482ae1301e315fe3